packages/flutter/lib/src/material/menu_bar.dart

```diff
@@ -0,0 +1,3256 @@
+    final List<RenderBox> renderBoxes = <RenderBox?>[
+      menuBarContext.findRenderObject() as RenderBox?,
+      ..._menuRenderBoxes,
+    ].where((RenderBox? box) => box != null).cast<RenderBox>().toList();
+    for (final RenderBox renderBox in renderBoxes) {
+      assert(renderBox.attached);
+      isInsideMenu =
+          renderBox.hitTest(BoxHitTestResult(), position: renderBox.globalToLocal(event.position)) || isInsideMenu;
+    }
+    if (!isInsideMenu) {
+      closeAll();
+    } else {
+      _focusBeforeClick = FocusManager.instance.primaryFocus;
+    }
+  }
+
+  /// Registers or updates the given menu in the menu controller whenever a menu
+  /// item widget is created or updated.
+  void registerMenu({
+    required BuildContext menuContext,
+    required _MenuNode node,
+    WidgetBuilder? menuBuilder,
+    EdgeInsets? menuPadding,
+    FocusNode? buttonFocus,
+  }) {
+    if (node.focusNode != buttonFocus) {
+      node.focusNode?.removeListener(_handleItemFocus);
+      node.focusNode = buttonFocus;
+      node.focusNode?.addListener(_handleItemFocus);
+      if (buttonFocus != null) {
+        _focusNodes[buttonFocus] = node;
+      }
+    }
+
+    node.menuPadding = menuPadding;
+    node.menuBuilder = menuBuilder;
+
+    if (node == _pendingFocusedMenu) {
+      node.focusNode?.requestFocus();
+      _pendingFocusedMenu = null;
+    }
+  }
+
+  /// Unregisters the given context from the menu controller.
+  ///
+  /// If the given context corresponds to the currently open menu, then close
+  /// it.
+  void unregisterMenu(_MenuNode node) {
+    node.focusNode?.removeListener(_handleItemFocus);
+    node.focusNode = null;
+    node.menuBuilder = null;
+    _focusNodes.remove(node.focusNode);
+    if (node == _pendingFocusedMenu) {
+      _pendingFocusedMenu = null;
+    }
+    if (openMenu == node) {
+      close(node);
+    }
+  }
+
+  // Used to register the menu's render box whenever it changes, so that it can
+  // be used to do hit detection and find out if a pointer event hit a menu or
+  // not without participating in the gesture arena.
+  void registerMenuRenderObject(RenderBox menu) {
+    _menuRenderBoxes.add(menu);
+  }
+
+  // Used to unregister the menu's previous render box whenever it changes, or
+  // remove it when it is disposed.
+  void unregisterMenuRenderObject(RenderBox menu) {
+    _menuRenderBoxes.remove(menu);
+  }
+
+  // Handles focus notifications for menu items so that the focused item can be
+  // set as the currently open menu.
+  void _handleItemFocus() {
+    if (openMenu == null) {
+      // Don't traverse the menu hierarchy on focus unless the user opened a
+      // menu already.
+      return;
+    }
+    final _MenuNode? focused = focusedItem;
+    if (focused != null && !isAnOpenMenu(focused)) {
+      openMenu = focused;
+    }
+  }
+
+  _MenuNode? get focusedItem {
+    final Iterable<FocusNode> focusedItems = _focusNodes.keys.where((FocusNode node) => node.hasFocus);
```

Comment:
can this just use `firstWhere`?

Reply:
I thought you could configure firstWhere to be a `firstOrNull` (see below). But thinking about this again, that probably will not work because `null` here is not of non-nullable Type `FocusNode`.

```dart
firstWhere((FocusNode node) => node.hasFocus, orElse: () => null)
```